source/non_matching/coupling.cc

```diff
@@ -40,6 +40,72 @@
 DEAL_II_NAMESPACE_OPEN
 namespace NonMatching
 {
+  namespace internal
+  {
+    template <int dim0, int dim1, int spacedim>
+    std::pair<std::vector<Point<spacedim>>, std::vector<unsigned int>>
+    compute_used_cells(const GridTools::Cache<dim0, spacedim> &cache,
+                       const DoFHandler<dim1, spacedim> &      immersed_dh,
+                       const Quadrature<dim1> &                quad,
+                       const Mapping<dim1, spacedim> &         immersed_mapping,
+                       const bool                              tria_is_parallel)
+    {
+      const auto &                 immersed_fe = immersed_dh.get_fe();
+      std::vector<Point<spacedim>> used_cell_ids;
+      // Keep track of which cells we actually used
+      std::vector<unsigned int> used_cells;
+      {
+        FEValues<dim1, spacedim> fe_v(immersed_mapping,
+                                      immersed_fe,
+                                      quad,
+                                      update_quadrature_points);
+        unsigned int             cell_id = 0;
+        for (const auto &cell : immersed_dh.active_cell_iterators())
+          {
+            bool use_cell = false;
+            if (tria_is_parallel)
+              {
+                const auto &bbox = cell->bounding_box();
+                std::vector<std::pair<
+                  BoundingBox<spacedim>,
+                  typename Triangulation<dim0, spacedim>::active_cell_iterator>>
+                  out_vals;
+                cache.get_cell_bounding_boxes_rtree().query(
+                  boost::geometry::index::intersects(bbox),
+                  std::back_inserter(out_vals));
+                // Each bounding box corresponds to an active cell
+                // of the imbedding triangulation: we now check if
+                // the current cell, of the imbedded triangulation,
+                // overlaps a locally owned cell of the imbedding one
+                for (const auto &bbox_it : out_vals)
+                  if (bbox_it.second->is_locally_owned())
+                    {
+                      use_cell = true;
+                      used_cells.emplace_back(cell_id);
+                      break;
+                    }
+              }
+
+            // If the triangulation is not parallel we're using all cells
+            if (!(tria_is_parallel) || use_cell)
```

Comment:
I found this hard to read. Maybe you can rewrite the code above as follows:
```
  if (tria_is_parallel)
    {... as above ...}
  else
    // for sequential triangulations, simply use all cells
    use_cell = true;
```
and then here simply do
```
  if (use_cell)
```